src/dctest/expressions.cljs

```diff
@@ -9,33 +9,44 @@
             [clojure.walk :refer [stringify-keys]]
             ["ebnf" :as ebnf]))
 
+(declare read-ast)
+
+(def supported-contexts
+  #{"env" "process"})
```

Comment:
I think we should keep this namespace generic and move this restriction up into dctest proper if we really want to restrict this and not leave it open.

Reply:
I did it here to keep it with the "standard library." Also if we move it out, I think that implies we have pass it in with things like read-ast to get validation results/errors.  Is that what you're thinking?

Continuing that train of thought, if we pull this out, I'd be interested to define counterparts in core (maybe?) that support the simpler `(read-ast text)` and then that could call into expressions to a function like `(read-ast supported text)` where supported (better name TBD) would include the supported contexts, standard library, and anything else we might need at read-time?  What do you think of this direction?

Reply:
(Note: if defined in core, we would only define the higher level things we'd need.  I don't know if we would actually need a `read-ast` equivalent per se, but if so, it would probably just be `read`.  We might only need `interpolated-text`, `read-eval` (maybe `read-eval-print` for completeness?) and something like `read-errors`)

Reply:
I guess I'm conceptualizing context as a runtime construct and not a syntactical construct and therefore all lookup errors would be runtime. Or another way of thinking about it is that context is essentially our lexical scope. So "foo.bar" is effectively the equivalent to "context.foo.bar" and therefore the same lookup and error handling logic for ".bar" should apply to ".foo" (at the point when/if we switch . to a checked property lookup.

Reply:
I'm totally onboard with the idea of contexts being runtime, but shouldn't the contexts we choose to expose be known ahead of time (and documented), thus able to be statically checked?  I really will feel bad for the unfortunate person who runs a 1 hr dctest suite and it fails for `proces.pid` being undefined :-P (I mean, even waiting 2 minutes to find out that information stinks)

Reply:
I guess I'm not convinced that `proces.platform` should be a syntax error at "parse" time while `process.platfom` would be a runtime issue (currently silent since it would be nil/null). They both feel like the same type of issue and treating "process" and "env" as part of syntax of expressions doesn't feel right to me. Also, I feel like at some point we may want a way to update a context in one step for subsequent steps (and between tests if they have a dependency relationship). Not 100% certain we'll want that, but making it syntax rather than runtime construct closes off that option. So I guess the answer to your question is that, at the expression layer of abstraction, I don't think contexts should be syntactical and known ahead of time.

This probably does show that it's important to change `.prop` to be a checked lookup (and `?.prop` is the uncheck version). At least then you get a clear error even if it's a late one.

Reply:
Ok, I've added ahead-of-time checking for `platform`, `pid`, etc. now ... just kidding ;-). I do prefer to check as much as we can, but I'm ok to concede it.  FWIW I'm having a little bit of trouble imagining a feature where user changes wouldn't go into their own top-level (known ahead-of-time) context name, so they can't shadow something we provide (basically how `env` works now.  `env` is 100% owned by the user, and `process` is 100% owned by us/dctest)... but I'm not totally confident, so yeah, ok to leave it open for now.

One quick question before I rework/revert the commit regarding context name errors:  how do you feel about the ahead-of-time function name/arity errors? (I hate to even ask, because I like these, too.)

Also, I'm definitely good with the `?.prop` direction.

Reply:
🤣 Yes, I think functions will probably want to be user defined and lifted out of core expressions too. But let's not attack that yet. This is still early development exploration and we aren't beholden yet to any specific interface contract.

One of the ideas I'm trying to be cognizant of in thinking through all this is that this expression logic may at some point be lifted out of dctest so that we can use it other places. A near-term one that comes to mind is dcmon where being able to have fuller expressions for the `finished` clauses as expressions would be immediately useful ("at least 2 of the 4 successful", "A but not B", "3 seconds passed since this fired", etc).

I just spent a little bit of "hammock time" (I wish I had a real hammock but alas) thinking about this. We could actually get both user-defined contexts (and functions) and ahead-of-time validation if we pass the context (and function set) to the reader/checker and not just to the evaluator. Additionally, if we did that we could evaluate properties all the way down and not just the top-level names. If we start pushing up against Turing Completeness and Greenspun's Law i.e. if we start wanting fuller language features like lexical scope, high order functions, chaining, state, etc, then doing full semantic checking ahead of time starts to become problematic. But we can cross the bridge when we get there (I think).

In the specific context of functions, I'm guessing that we may want a couple of ways to define functions. With full user-specified metadata, and as plain functions. Note that in JS you can look at the "length" property of a function to determine the minimum number of arguments that the function requires. Native ClojureScript functions somewhat comply with this rule but there are odd cases such as `(.-length (constantly true)) -> 1` when it would be more accurate to return 0, and `(.-length (fn [a & b] (prn a b))) -> 2` when 1 would be more consistent with JS. Anyways, some food for thought.